### ISSUE TITLE
LSTM Operation support

### ISSUE THREAD

**Issue by tforgaard** [PINTO0309/onnx2tf#198]
### Issue Type

Feature Request

### onnx2tf version number

1.7.3

### onnx version number

1.11.0

### tensorflow version number

2.10.0

### Download URL for ONNX

Simple [lstm module](https://drive.google.com/file/d/1ZbvZEynKwdkEbs-jrTyh_Ozmrm3fHNpz/view?usp=sharing)
<img width="378" alt="Skjermbilde 2023-02-22 kl  11 11 41" src="https://user-images.githubusercontent.com/14992984/220596809-e4b61709-9bb9-42bf-855a-52a9912de10a.png">

### Parameter Replacement JSON

```json
nil
```

### Description

1. Purpose: Research project, training an RL agent in simulation, and then deploying the agent in the real world. I am trying to convert an RL policy all the way from a pytorch model to an edgeTPU compatible model to run realtime inference. The conversion process is pytorch model -> onnx -> TF -> TFLite -> EdgeTPU. The use of an RNN model is crucial for the project's success.

2. So far I have successfully transferred a policy without an LSTM module using [onnx-tensorflow](https://github.com/onnx/onnx-tensorflow), from PyTorch to edgeTPU. And I assume I would be able to do the same using onnx2tf. However, I have had no success when trying to convert a policy using an LSTM using onnx-tensorflow. As the repo is not actively maintained, I became interested in your work.

3. As I had no success using onnx-tensorflow, I instead resolved to create my own script for copying the weights for only the LSTM module from a pytorch model to a tensorflow model, skipping the onnx intermediate step, with success.

Now, this method requires me to cut out the LSTM module from my entire model and run my own conversion script on this module and glue together the network afterward. 

4. My hacky way of converting the LSTM module is obviously a tedious process, and being able to use onnx2tf for the whole conversion process would be extremely time-saving.

I see that you have given the LSTM op a couple of tries and was wondering if you could elaborate on what problems you faced when trying to implement the LSTM op? If possible I would like to try to help.

5. Copying weights from TF to pytorch (inverse of what I did): https://www.adrian.idv.hk/2020-12-31-torch2tf/

**Comment by PINTO0309** [comment]
First, see the results of my two weeks of worrying and failed implementation.

https://github.com/PINTO0309/onnx2tf/tree/main/onnx2tf/ops
![image](https://user-images.githubusercontent.com/33194443/220599291-70eb5397-984f-4278-b9f8-0226d826c90b.png)


I am overflowing with the desire to implement `LSTM` myself.  I am almost alone and at a loss to come up with good ideas for implementation.

**Comment by PINTO0309** [comment]
Memorandum of Understanding. Not yet tested and implementation will be verified in the future.

- LSTM
  https://zenn.dev/pinto0309/scraps/430cea62b1eb9d

**Comment by PINTO0309** [comment]
It can be converted when Bias initialization is not required, but the size of the tensor is unmatched when Bias initialization is required.

https://github.com/PINTO0309/onnx2tf/blob/exp_lstm/onnx2tf/ops/LSTM.py

**Comment by PINTO0309** [comment]
The conversion of `LSTM` with `Bias` was successful after successful initialization of `Bias`. However, I don't know how to handle `initial_state`, so I need to find out how to use `initial_state`.

https://github.com/PINTO0309/onnx2tf/blob/imp_lstm/onnx2tf/ops/LSTM.py

![image](https://user-images.githubusercontent.com/33194443/229653173-ac22e21e-e67c-46be-9d08-e70b66180fe2.png)

![image](https://user-images.githubusercontent.com/33194443/229653148-0788b649-812d-4670-858d-6081a1228592.png)

**Comment by PINTO0309** [comment]
- `LSTM`
  - Major revision of `LSTM` conversion process.
  - Supports arbitrary activation functions, biases and initial_state.
  - However, it does not yet take `P` into account.
  - Also, there is an error (`Unmatched`) in the inference results between ONNX and TFLite, although the cause is not entirely clear.
  - [LSTM.tanh.bidirectional_b10.onnx.zip](https://github.com/PINTO0309/onnx2tf/files/11154879/LSTM.tanh.bidirectional_b10.onnx.zip)
  - onnx
    ![image](https://user-images.githubusercontent.com/33194443/229973614-da561247-98cf-40ad-a221-0e223d53f95c.png)
  - tflite
    ![image](https://user-images.githubusercontent.com/33194443/229973407-758066b0-9a25-4706-979e-559d7fec5b63.png)

![image](https://user-images.githubusercontent.com/33194443/229974613-d61b9b3b-f58a-4ba8-aa73-221829178a6f.png)

**Comment by PINTO0309** [comment]
- `LSTM`
  - Fixed `LSTM` final output to be inverted on the axis of the sequence when `direction == "reverse"`.
  - For `LSTM` that is not `direction == "bidirectional"`, the converted output now matches ONNX exactly.
  - https://github.com/PINTO0309/onnx2tf/files/11154879/LSTM.tanh.bidirectional_b10.onnx.zip
  - Ref: https://github.com/tensorflow/tensorflow/issues/40345
  - command
    ```
    onnx2tf -i LSTM.tanh.bidirectional_b10.onnx -kat Input3
    ```
  - onnx (forward + reverse)
    ![image](https://user-images.githubusercontent.com/33194443/229992326-4d994b0c-131f-4bea-bcc0-8783e9f1c9a4.png)
  - tflite
    ![image](https://user-images.githubusercontent.com/33194443/229991459-3315222d-f05b-4f10-819c-fecb5a8748d8.png)
  - result
    ![image](https://user-images.githubusercontent.com/33194443/229991528-19a26bc2-ca37-4637-be98-306902c4d02d.png)

**Comment by PINTO0309** [comment]
- `LSTM`
  - Fixed to calculate after summing the `LSTM` Bias for `W` and `R`.
  - `direction == bidirectional` is now converted correctly.
  - https://s3.ap-northeast-2.wasabisys.com/temp-models/onnx2tf_198/text_recognition_CRNN_EN_2021sep.onnx
  - onnx
    ![image](https://user-images.githubusercontent.com/33194443/230241791-2c4809da-6ddf-4b0f-997d-e8c7dcd36194.png)
  - tflite
    ![image](https://user-images.githubusercontent.com/33194443/230241863-98b1e06d-8403-47f8-a99f-7c3f66459d8a.png)
  - result
    ![image](https://user-images.githubusercontent.com/33194443/230241916-27e3e757-e191-45ef-b482-8af03929a3ce.png)
- `Squeeze`
  - Parameter replacement process was omitted from the implementation, so it was added.

**Comment by PINTO0309** [comment]
@tforgaard

You may not be interested anymore, but I have addressed the `LSTM` conversion. `v1.8.17`
Therefore, close this issue.

- replace_lstm.json
  ```json
  {
    "model": "https://s3.ap-northeast-2.wasabisys.com/temp-models/onnx2tf_198/torch_lstm.onnx",
    "issue": "https://github.com/PINTO0309/onnx2tf/issues/198",
    "command": "onnx2tf -i torch_lstm.onnx -prf replace_lstm.json -kat input h_in c_in",
    "format_version": 1,
    "operations": [
      {
        "op_name": "/Squeeze",
        "param_target": "attributes",
        "param_name": "axes",
        "values": [1]
      },
      {
        "op_name": "/Transpose_1",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [0,1,2]
      }
    ]
  }
  ```
- command
  ```
  onnx2tf -i torch_lstm.onnx -prf replace_lstm.json -kat input h_in c_in
  ```
- results
  ![image](https://user-images.githubusercontent.com/33194443/230247125-2557e9cd-777d-46ad-b81c-3bf821a89130.png)
  ![image](https://user-images.githubusercontent.com/33194443/230246820-b77b9ab6-e0f1-4577-a856-ab01f7e04ec6.png)

**Comment by tforgaard** [comment]
Thanks @PINTO0309,

I am very grateful for the work you have done, I hope it will benefit others as well!
